### PR TITLE
Cloud Authorization

### PR DESCRIPTION
**Pull request by VJalili**
This PR implements the logic of accessing a user's cloud-based resources without asking for their credentials, leveraging [CloudAuthz](https://github.com/galaxyproject/cloudauthz). In general, this PR implements the following: 

1. defines a `cloudauthz` class in Galaxy model;
2. implements migration script to create a table in Galaxy database to store instances of the new class;
3. updates mappings by required joins;
4. creates API and manager so that user can define a `cloudauthz` instance;
5. integrates [`cloudauthz`](https://github.com/galaxyproject/cloudauthz) with `AuthnzManager`;
6. replaces `credentials` with a `cloudauthz` ID in the  cloud `upload` API;
7. updates `CloudManager` to leverage `cloudauthz` to get credentials.
8. updates `CloudManager` to adhere with `v1.0.0` cloudbridge interface.  

This PR requires the following to updated: 
- [x] Version 1.0.0 of CloudBridge; #6660 
- [x] Update [`cloudauthz`](https://github.com/galaxyproject/cloudauthz) to resolve dependency conflicts with Galaxy;
- [x] Some of the raised exceptions cause 500 error (e.g., exceptions in `galaxy/api/cloudauthz.py`).

### CODE REVIEW

**Comment by VJalili** [comment]
@galaxybot test this

**Comment by jmchilton** [comment]
Thanks! I have pretty much the same comment as here https://github.com/galaxyproject/galaxy/pull/5835#issuecomment-382505057.

**Comment by martenson** [comment]
I am pushing this back to 18.09 since we probably won't have enough time to test is and there are unaddressed comments.

@VJalili please let me know if this complicates things significantly

**Comment by VJalili** [comment]
@jmchilton Thanks for reviewing this. The exception handling is now updated, and it uses `galaxy.exceptions` as you recommended.

**Comment by VJalili** [comment]
@galaxybot test this

**Comment by VJalili** [comment]
@galaxybot test this

**Comment by VJalili** [comment]
@galaxybot test this

**Comment by VJalili** [comment]
@afgane Thanks!

**Comment by afgane** [comment]
The high level usage model here is as follows:
1. Head to https://console.cloud.google.com/apis/credentials and create a OAuth 2.0 client ID
2. Enter those values into `config/oidc_backends_config.xml`
3. Head to AWS IAM dashboard, create a role with desired policy (e.g. `AmazonS3ReadOnlyAccess`), and attach the equivalent of the trust relationship included below. Note the role ARN.
4. Start Galaxy and GET http://127.0.0.1:8080/authnz; note down the `id`
5. Create a new authz entry by issuing a POST with the content included below
6. Test that it all works by uploading a file from a private bucket by issuing another POST request (see body below)

- Trust relationship (where the value for `accounts.google.com:aud` is the respective `client_id` from the Google OAuth client registration):
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "accounts.google.com"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "accounts.google.com:aud": "893677542423-4t761afe33k3o9mu56u6p6p1ctk8o89f.apps.googleusercontent.com"
        }
      }
    }
  ]
}
```

- Create authz entry body:
```
{
  "provider": "aws",
  "authn_id":"f2db41e1fa331b3e",
  "config": {
  	"role_arn": "arn:aws:iam::861460482541:role/s3-read-access"
  }
}
```

- Upload data from S3
```
{
  "history_id": "f597429621d6eb2b",
  "bucket": "cm-log",
  "objects": ["cm_log2011-01-13-17-20-31-213597BDDB2013BE", "cm_log2011-01-13-17-20-46-1D133AFF95259DFE"],
  "authz_id": "f2db41e1fa331b3e"
}
```

**Comment by martenson** [comment]
@afgane ideally this would go to docs I think

**Comment by afgane** [comment]
It will, with a lot more detail, but for now these are the cliff notes.

File: lib/galaxy/webapps/galaxy/api/cloudauthz.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        rtv = []
+        for cloudauthz in trans.user.cloudauthz:
+            rtv.append(self.cloudauthz_serializer.serialize_to_view(
+                cloudauthz, user=trans.user, trans=trans, **self._parse_serialization_params(kwargs, 'summary')))
+        return rtv
+
+    @expose_api
+    def create(self, trans, payload, **kwargs):
+        """
+        * POST /api/cloud/authz/create
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/api/cloudauthz.py:62]: Probably this should be a POST to ``/api/cloud/authz`` instead?

File: lib/galaxy/webapps/galaxy/buildapp.py [hunk 2]
```diff
@@ -295,6 +295,9 @@ def populate_api_routes(webapp, app):
     webapp.mapper.resource('group', 'groups', path_prefix='/api')
     webapp.mapper.resource_with_deleted('quota', 'quotas', path_prefix='/api')
 
+    webapp.mapper.connect('/api/cloud/authz/', action='index', controller='cloudauthz')
+    webapp.mapper.connect('/api/cloud/authz/create', action='create', controller='cloudauthz')
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/buildapp.py:299]: So here you can specify actions, specify the one above to be a GET and this one to be a POST. There are examples in this file.

Comment (VJalili) [reply lib/galaxy/webapps/galaxy/buildapp.py:299]: Please check the following PR and let me know if that address your point: https://github.com/galaxyproject/galaxy/pull/6876

**Pull request merged**